src/v/cluster/rm_stm.cc

```diff
@@ -2605,4 +2636,79 @@ ss::future<> rm_stm::handle_eviction() {
       });
 }
 
+void rm_stm::spawn_background_clean_for_pids(rm_stm::clear_type type) {
+    switch (type) {
+    case rm_stm::clear_type::tx_pids: {
+        if (_log_state.fence_pid_epoch.size() <= _max_saved_pids_count()) {
+            return;
+        }
+        break;
+    }
+    case rm_stm::clear_type::idempotent_pids: {
+        if (_log_state.seq_table.size() <= _max_saved_pids_count()) {
+            return;
+        }
+        break;
+    }
+    }
+
+    ssx::spawn_with_gate(_gate, [this, type] {
+        return _state_lock.hold_read_lock().then(
+          [this,
+           type](ss::basic_rwlock<>::holder unit) mutable -> ss::future<> {
+              switch (type) {
+              case rm_stm::clear_type::tx_pids: {
+                  vlog(_ctx_log.trace, "Gonna clear old tx pids");
+                  co_await clear_old_tx_pids();
+                  break;
+              }
+              case rm_stm::clear_type::idempotent_pids: {
+                  vlog(_ctx_log.trace, "Gonna clear old idempotent pids");
+                  co_await clear_old_idempotent_pids();
+                  break;
+              }
+              }
+          });
+    });
+}
+
+ss::future<> rm_stm::clear_old_tx_pids() {
+    if (_log_state.fence_pid_epoch.size() < _max_saved_pids_count()) {
+        co_return;
+    }
+
+    std::vector<model::producer_identity> pids_for_delete;
+    for (auto [id, epoch] : _log_state.fence_pid_epoch) {
+        auto pid = model::producer_identity(id, epoch);
+        // If pid is not inside tx_seqs it means we do not have
+        // transaction for it right now
+        if (!_log_state.tx_seqs.contains(pid)) {
+            pids_for_delete.push_back(pid);
+        }
+    }
+
+    co_await ss::max_concurrent_for_each(
+      pids_for_delete, 32, [this](auto pid) -> ss::future<> {
+          // We have transaction for this pid
+          if (_log_state.tx_seqs.contains(pid)) {
+              return ss::now();
+          }
+          _mem_state.forget(pid);
+          _log_state.forget(pid);
+          return ss::now();
+      });
+}
+
+ss::future<> rm_stm::clear_old_idempotent_pids() {
+    while (_log_state.idempotent_pids_order.size() > _max_saved_pids_count()) {
```

Comment:
`.size()` is O(n). Maybe we need support `size_t size` near idempotent_pids_order

Reply:
you can use: `counted_intrusive_list` from `src/v/utils/intrusive_list_helpers.h`

Reply:
it provides `O(1)` `size()` method